### ISSUE TITLE
Изменение свойства из иерархии классов

### ISSUE THREAD

**Issue by tsouvarev** [tsouvarev/semantic_objects#16]
Проблема заключается в том, что если прикладной программист захочет поменять свойство, которое принадлежит не объекту, а классу, то непонятно по какому пути идти: 
1. Добавляется триплет, закрепляющий за этим объектом значение свойства 
2. Перевести объект в соответствующий класс 
3. Разрешить изменять только свойства, инициализированные для объектов

То есть, если есть объект ChiantiClassico, который (например) унаследовал от WhiteWine hasColor = White, то следует ли переназначать цвет и если да, то где?

**Comment by v--victoria** [comment]
Ну в моем понимании, ситуация, когда у класса есть инициализированное свойство (есть триплет: класс - свойство - значение_свойства), это что-то типа статического поля класса в ООП(то есть общее поле для всех объектов класса). 
Поэтому раз уж такое поле есть, я бы разрешила его менять, но только в классе (т.е. менять именно триплет класс - свойство - значение_свойства, но не объект_класса - свойство - новое_значение).
Мне кажется, так логичнее (это, видимо, вариант 2).

Если менять его в объекте (вариант 1), наверно, вообще получится путаница и каша.

Вариант 3 - в принципе тоже логичный, но вдруг понадобится изменить значение такой штуки в нашей онтологии мудла.

**Comment by tsouvarev** [comment]
да, но если менять в классе, то свойство поменяется для всех объектов этого класса. та же путаница может получиться)

**Comment by v--victoria** [comment]
Да, поменяется для всех объектов. Как в этом случае может быть путаница?

**Comment by tsouvarev** [comment]
Просто это неочевидное поведение. Меняешь вроде свойство для объекта (программист же не в курсе, где это свойство определено - для объекта или для класса; разве что ручками в онтологию лезть и проверять), а оно меняется для всего класса. 
Напомню, что в объектной модели питона если свойство принадлежало родительскому классу и меняется у объекта, то оно сохраняется в объекте - то есть, это равносильно введению триплета (объект, свойство, новое значение)

**Comment by v--victoria** [comment]
Да, согласна. Тогда 2 вариант, пожалуй, самый неудобный (тут как будто программист совершает очевидную ошибку).
Надо подумать, будут ли у нас ситуации, когда может потребоваться вариант 1 (хотя наверняка могут быть). Если нет, то тогда самый строгий вариант 3.

**Comment by tsouvarev** [comment]
В тред призывается научный руководитель))))
А если серьезно, то все варианты плохие:
1. Допустим, добавили триплет. Сразу придется сказать, что значения, определенные для объекта, имеют привилегию над значениями из класса. Через неделю меняем значение в классе, ожидая, что оно наследуется всеми объектами. Ой.
2. Переводить в другой класс с одной стороны весело и клево, с другой стороны муторно и тяжело технически. Может не существовать подходящего класса, программист может не ожидать такой вольности и т.д. В данном случае, библиотека слишком многое решает за программиста
3. Вроде бы тоже адекватное решение, но как объяснить программисту, что в этом классе можно менять свойство, потому как оно определено у объекта, а в этом нельзя, потому что свойство из класса; при этом, объекты могут быть похожи как два близнеца и свойства даже могут называться одинаково. В этом случае либо программа может упасть в любой момент, либо накручивать кучу проверок.
Но за неимением лучшего я склоняюсь к 1 или 3 как наиболее простым.

**Comment by v--victoria** [comment]
В случае 1 программист всего лишь должен понимать, что изменяемое значение свойства в классе, возможно, не приведет к изменению его же в объекте. Это и не так страшно, наверно, и это должно быть уже на его совести (то есть на моей)=)
В 3 случае программисту можно возвращать код ошибки. С похожими объектами сложнее.

**Comment by tsouvarev** [comment]
Стало быть, останавливаемся на варианте №1:
Добавляется триплет, закрепляющий за этим объектом значение свойства

**Comment by v--victoria** [comment]
Да, ну и по питоновски получится.
